Source/Core/Core/AchievementManager.cpp

```diff
@@ -1016,694 +520,461 @@ void AchievementManager::FilereaderClose(void* file_handle)
+                                               .m_unlocked_badges[client_event->achievement->id]
+                                               .badge) :
                       nullptr);
-  if (m_game_data.achievements[game_data_index].category == RC_ACHIEVEMENT_CATEGORY_CORE)
-  {
-    m_queue.EmplaceItem([this, event_id] { AwardAchievement(event_id); });
-    PointSpread spread = TallyScore();
-    if (spread.hard_points == spread.total_points &&
-        it->second.remote_unlock_status != UnlockStatus::UnlockType::HARDCORE)
-    {
-      OSD::AddMessage(
-          fmt::format("Congratulations! {} has mastered {}", m_display_name, m_game_data.title),
-          OSD::Duration::VERY_LONG, OSD::Color::YELLOW,
-          (Config::Get(Config::RA_BADGES_ENABLED)) ? DecodeBadgeToOSDIcon(m_game_badge.badge) :
-                                                     nullptr);
-    }
-    else if (spread.hard_points + spread.soft_points == spread.total_points &&
-             it->second.remote_unlock_status == UnlockStatus::UnlockType::LOCKED)
-    {
-      OSD::AddMessage(
-          fmt::format("Congratulations! {} has completed {}", m_display_name, m_game_data.title),
-          OSD::Duration::VERY_LONG, OSD::Color::CYAN,
-          (Config::Get(Config::RA_BADGES_ENABLED)) ? DecodeBadgeToOSDIcon(m_game_badge.badge) :
-                                                     nullptr);
-    }
-  }
-  ActivateDeactivateAchievement(event_id, Config::Get(Config::RA_ACHIEVEMENTS_ENABLED),
-                                Config::Get(Config::RA_UNOFFICIAL_ENABLED),
-                                Config::Get(Config::RA_ENCORE_ENABLED));
 }
 
-void AchievementManager::HandleAchievementProgressUpdatedEvent(
-    const rc_runtime_event_t* runtime_event)
+void AchievementManager::HandleGameCompletedEvent(const rc_client_event_t* client_event,
+                                                  rc_client_t* client)
 {
-  if (!Config::Get(Config::RA_PROGRESS_ENABLED))
-    return;
-  auto it = m_unlock_map.find(runtime_event->id);
-  if (it == m_unlock_map.end())
-  {
-    ERROR_LOG_FMT(ACHIEVEMENTS, "Invalid achievement progress updated event with id {}.",
-                  runtime_event->id);
-    return;
-  }
-  AchievementId game_data_index = it->second.game_data_index;
-  FormattedValue value{};
-  if (rc_runtime_format_achievement_measured(&m_runtime, runtime_event->id, value.data(),
-                                             FORMAT_SIZE) == 0)
-  {
-    ERROR_LOG_FMT(ACHIEVEMENTS, "Failed to format measured data {}.", value.data());
-    return;
-  }
-  OSD::AddMessage(
-      fmt::format("{} {}", m_game_data.achievements[game_data_index].title, value.data()),
-      OSD::Duration::SHORT, OSD::Color::GREEN,
-      (Config::Get(Config::RA_BADGES_ENABLED)) ?
-          DecodeBadgeToOSDIcon(it->second.unlocked_badge.badge) :
-          nullptr);
+  auto* user_info = rc_client_get_user_info(client);
```

Comment:
Could `user_info` or `game_info` ever return null?

Reply:
They return null if there is no user/game currently loaded, yes.

Reply:
Ah ok.  Seems unlikely to be an issue if we're getting an event but can't hurt to be safe:

```
auto* user_info = rc_client_get_user_info(client);
if (!user_info) [[unlikely]]
  return;
auto* game_info = rc_client_get_game_info(client);
if (!game_info) [[unlikely]]
  return;
bool hardcore = rc_client_get_hardcore_enabled(client);
  ```